File: lua/examples/guards/scripts/GuardChief.lua
---@uuid c69ae5fa-de26-4ee5-b70c-113df285f6e2
---@class GuardChief : Script
---@field gaurd_prefab Prefab
---@field initial_count number
---@field private initialized boolean
---@field private frame_skipped_for_beacons boolean
GuardChief = script_class()


function GuardChief:on_update(ctx)
    if not self.frame_skipped_for_beacons then
        self.frame_skipped_for_beacons = true;
        return;
    end
    if not self.initialized then
        self.initialized = true;
        for i = 1, self.initial_count do
            local beacons = Plugin:get("Game").beacons
            if #beacons > 0 then
                local position = beacons[math.random(#beacons)]

                local orientation = Quaternion:from_axis_angle(Vector3.Y, math.random() * 2 * math.pi)

                local guard = self.gaurd_prefab:instantiate_at(
                    position,
                    orientation
                )
                guard:find_script("Guard"):init(i)
                Log:info(string.format("guard spawned at %s", tostring(position)))
            else
                Log:err("cannot spawn guards: no beacons found")
            end
        end
    end
end

File: cs/examples/Guards/GuardChief.cs
using FyroxLite;

[Uuid("c69ae5fa-de26-4ee5-b70c-113df285f6e2")]
public class GuardChief : NodeScript
{
    private Prefab gaurd_prefab;
    private float initial_count;

    [HideInInspector] [Transient] private bool initialized;

    [HideInInspector] [Transient] private bool frame_skipped_for_beacons;

    protected override void OnUpdate(float dt)
    {
        if (!frame_skipped_for_beacons)
        {
            frame_skipped_for_beacons = true;
            return;
        }

        if (!initialized)
        {
            initialized = true;
            for (int i = 1; i <= initial_count; i++)
            {
                var beacons = Plugin.Get<Game>().beacons;
                if (beacons.Count > 0)
                {
                    var position = beacons[new Random().Next(beacons.Count)];
                    var angle = (float)(new Random().NextDouble() * 2 * Math.PI);

                    var guard = gaurd_prefab.InstantiateAt(position, new Quaternion(Vector3.Up, angle));
                    guard.FindScript<Guard>().Init(i);

                    Log.Info($"Guard spawned at {position}");
                }
                else
                {
                    Log.Err("Cannot spawn guards: no beacons found");
                }
            }
        }
    }
}